crates/miden-objects/src/account/component/template/storage/toml.rs

```diff
@@ -385,6 +393,84 @@ impl<'de> Deserialize<'de> for StorageEntry {
+
+impl InitStorageData {
+    /// Creates an instance of [`InitStorageData`] from a TOML string.
+    ///
+    /// This method parses the provided TOML and flattens nested tables into
+    /// dot‑separated keys using [`StorageValueName`] as keys. All values are converted to plain
+    /// strings (so that, for example, `key = 10` and `key = "10"` both yield
+    /// `String::from("10")` as the value).
+    ///
+    /// # Errors
+    ///
+    /// - If duplicate keys or empty tables are found in the string
+    /// - If the TOML string includes arrays
+    pub fn from_toml(toml_str: &str) -> Result<Self, InitStorageDataError> {
+        let value: toml::Value = toml::from_str(toml_str)?;
+        let mut placeholders = BTreeMap::new();
+        // Start with an empty prefix (i.e. the default, which is an empty string)
+        Self::flatten_parse_toml_value(StorageValueName::default(), &value, &mut placeholders)?;
+        Ok(InitStorageData::new(placeholders))
+    }
+
+    /// Recursively flattens a TOML `Value` into a flat mapping.
+    ///
+    /// When recursing into nested tables, keys are combined using
+    /// [`StorageValueName::with_suffix`]. If an encountered table is empty (and not the top-level),
+    /// an error is returned. Arrays are not supported.
+    fn flatten_parse_toml_value(
+        prefix: StorageValueName,
+        value: &toml::Value,
+        map: &mut BTreeMap<StorageValueName, String>,
+    ) -> Result<(), InitStorageDataError> {
+        match value {
+            toml::Value::Table(table) => {
+                // If this is not the root and the table is empty, error
+                if !prefix.as_str().is_empty() && table.is_empty() {
+                    return Err(InitStorageDataError::EmptyTable(prefix.as_str().into()));
+                }
+                for (key, val) in table {
+                    // Create a new key and combine it with the current prefix.
+                    let new_key = StorageValueName::new(key.to_string())
+                        .map_err(InitStorageDataError::InvalidStorageValueName)?;
+                    let new_prefix = prefix.clone().with_suffix(&new_key);
+                    Self::flatten_parse_toml_value(new_prefix, val, map)?;
+                }
+            },
+            toml::Value::Array(_) => {
+                return Err(InitStorageDataError::ArraysNotSupported);
+            },
+            toml_value => {
+                // Get the string value, or convert to string if it's some other type
+                let value = match toml_value {
+                    toml::Value::String(s) => s.clone(),
+                    _ => value.to_string(),
+                };
+                map.insert(prefix, value);
+            },
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum InitStorageDataError {
+    #[error("failed to parse TOML: {0}")]
+    InvalidToml(#[from] toml::de::Error),
```

Comment:
`#[from]` implies `#[source]` so this both includes the error in the display impl as well as return it from source. It should only do one of those two. For simplicity, you could just remove the `: {0}`.